Add GH action: convert yaml to json

radishmouse opened the pull request:
Closes #3
Closes #8 

This PR:
* Runs the yaml to json conversion
* Validates existing json and any new json produced by the conversion
* If all json is valid, commit new json to repo

Changes to the `validate` workflow were necessary since it is called from the `convert` workflow:
* Makes the `validate` workflow reusable, but retains standalone functionality
* Adds inputs and outputs to `validate` 
  * the `should_download` input tells `validate` to download the JSON artifacts produced during the GH action
  * the `is_valid` output contains the status of the validation (whether all json files were valid)
* Per feedback, a python script now wraps the JSON validation command line tool

Tested locally using [nektos/act](https://github.com/nektos/act):
* `act pull_request` simulates a PR
  * This tests the `validate` workflow in standalone mode
* `act push --artifact-server-path /tmp` simulates a push
  * This is for testing the `convert` workflow (which calls `validate` before attempting to commit)

Note: only partially tested on GitHub. (This PR includes an example JSON file that passes validation.) #13 has been created to address this.

malini commented on .github/workflows/convert-yaml-to-json.yml:
```diff
@@ -0,0 +1,26 @@
+name: convert-yaml-to-json
+
+on:
+  [push]
```
Question, where will the pushed yaml live?

we should run this action only when we see yaml being pushed up, so maybe we can have a yaml/ folder at root so people push their files there. Then we can have this action run against changes on that folder like so:
```
on:
  push:
    paths:
      - 'yaml/**.yaml'

```
see [this page](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore) for more details

radishmouse replied:
I think that's a great idea!

(I don't think we had decided on a folder for the pushed yaml, so `yaml` in the root is :+1: )

radishmouse commented:
One thing that's missing is that the resulting json files need to be commited/pushed to the repo.
I'll update the GH action to include those steps

malini commented on .github/workflows/convert-yaml-to-json.yml:
```diff
@@ -0,0 +1,24 @@
+          pip install -r requirements.txt
+      - name: Run python script
+        run: |
+          python convert-yaml.py test/data/yaml test/data/json
```
shouldn't this action run against the new yaml folder?

radishmouse commented:
@malini I have tested this code heavily locally, but not on GitHub.

I have created #13 for testing on GH, but would feel more comfortable with at least one more set of eyes on the workflow code before I merge.

MelissaAutumn commented on .github/workflows/convert.yml:
```diff
@@ -0,0 +1,71 @@
+        if: ${{ env.NO_COMMIT != 'true' }}
+        run: |
+          git config --global user.name "github-actions"
+          git config --global user.email "github-actions@github.com"
```
We might want to use a branded account here? For our website builds we use https://github.com/thundernest-bot

If you tie the email to a secret we can probably re-use that. Thoughts @Sancus ?

MelissaAutumn commented on .github/workflows/convert.yml:
```diff
@@ -0,0 +1,71 @@
+name: convert-yaml-to-json
+
+on:
+  push:
```
Do we want this to be limited to specific branches to avoid over-commiting PRs?

aaspinwall replied:
How about on push to main and on pull requests?

malini replied:
@aaspinwall's suggestion sounds good to me!

MelissaAutumn reviewed (changes requested):
A few things! I haven't tested the actual flow yet but it looks like it should be okay?  I guess it technically should run on this PR 😄

radishmouse commented on .github/workflows/convert.yml:
```diff
@@ -0,0 +1,71 @@
+      - name: Run python script
+        run: |
+          if [ -d "yaml" ] && [ "$(ls -A yaml/*.yaml 2>/dev/null)" ]; then
+            python convert-yaml.py yaml json
```
Ah! In that case, do you think we should wrap the validation command (`jv schema.json json/*.json`) in a shell script that handles failures similarly?

That would give us a bit of symmetry between the validation and conversion commands.

MelissaAutumn replied:
You could even wrap it in a python script with subprocess.popen if you want to make the language consistent.

radishmouse replied:
haha GH yaml that runs a shell command that runs a python script that runs a command line tool written in go.

but also, I like this suggestion - I'll go with the python wrapper! :+1:

radishmouse commented:
@MelissaAutumn I've added a python wrapper script for validation and moved the checks for yaml & json files from the workflow file to python.
When you get a chance, please take a look?

MelissaAutumn commented on .github/workflows/validate.yml:
```diff
@@ -1,27 +1,56 @@
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+          # cache: "pip"
```
You can re-enable this, and install deps
```
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install -r requirements.txt
```
Technically this is a separate system, and you're running python so you should probably have the deps installed!

radishmouse replied:
Thanks for this!

MelissaAutumn reviewed (approved):
One change but it looks like it should be good?
